plugins/modules/s3_object.py

```diff
@@ -969,287 +1200,84 @@ def main():
-                create_bucket(module, s3, bucket, location)
-                if not acl_disabled:
-                    # only use valid object acls for the create_dirkey function
-                    module.params['permission'] = object_acl
-                create_dirkey(module, s3, bucket, dirobj, encrypt, expiry)
-
-    # Support for grabbing the time-expired URL for an object in S3/Walrus.
-    if mode == 'geturl':
-        if not bucket and not obj:
-            module.fail_json(msg="Bucket and Object parameters must be set")
-
-        keyrtn = key_check(module, s3, bucket, obj, version=version, validate=validate)
-        if keyrtn:
-            tags = get_current_object_tags_dict(s3, bucket, obj, version=version)
-            get_download_url(module, s3, bucket, obj, expiry, tags)
-        else:
-            module.fail_json(msg="Key %s does not exist." % obj)
-
-    if mode == 'getstr':
-        if bucket and obj:
-            keyrtn = key_check(module, s3, bucket, obj, version=version, validate=validate)
-            if keyrtn:
-                try:
-                    download_s3str(module, s3, bucket, obj, version=version)
-                except Sigv4Required:
-                    s3 = get_s3_connection(module, aws_connect_kwargs, location, ceph, endpoint_url, sig_4=True)
-                    download_s3str(module, s3, bucket, obj, version=version)
-            elif version is not None:
-                module.fail_json(msg="Key %s with version id %s does not exist." % (obj, version))
-            else:
-                module.fail_json(msg="Key %s does not exist." % obj)
-
-    if mode == 'copy':
-        # if copying an object in a bucket yet to be created, acls for the bucket and/or the object may be specified
-        # these were separated into the variables bucket_acl and object_acl above
-        d_etag = None
-        if bucketrtn:
-            d_etag = get_etag(s3, bucket, obj)
+    func_mapping = {
+        'get': s3_object_do_get,
+        'put': s3_object_do_put,
+        'delobj': s3_object_do_delobj,
+        'delete': s3_object_do_delete,
+        'list': s3_object_do_list,
+        'create': s3_object_do_create,
+        'geturl': s3_object_do_geturl,
+        'getstr': s3_object_do_getstr,
+        'copy': s3_object_do_copy
+    }
+    func = func_mapping[s3_object_params['mode']]
+    try:
+        func(module, s3, s3_object_params)
+    except botocore.exceptions.EndpointConnectionError as e:  # pylint: disable=duplicate-except
+        module.fail_json_aws(e, msg="Invalid endpoint provided")
+    except is_boto3_error_code(IGNORE_S3_DROP_IN_EXCEPTIONS):
+        module.warn(
+            "PutObjectAcl is not implemented by your storage provider. Set the permissions parameters to the empty list to avoid this warning"
+        )
+    except is_boto3_error_message('require AWS Signature Version 4'):  # pylint: disable=duplicate-except
+        raise Sigv4Required()
```

Comment:
This was is supposed to be caught early. If it happens at this level, it's a bug.